lib/model/folder_sendrecv.go

```diff
@@ -1334,16 +1294,85 @@ func (f *sendReceiveFolder) copierRoutine(in <-chan copyBlocksState, pullChan ch
 				state.copyDone(block)
 			}
 		}
-		if file != nil {
-			// os.File used to return invalid argument if nil.
-			// fs.File panics as it's an interface.
-			file.Close()
-		}
 
 		out <- state.sharedPullerState
 	}
 }
 
+// copyByWeakHash copies all blocks that can be found from their weak hash.
+// It returns an updated state with these blocks removed.
+func (f *sendReceiveFolder) copyByWeakHash(state copyBlocksState, w io.WriterAt) (copyBlocksState, error) {
+	file, err := f.fs.Open(state.file.Name)
+	if err != nil {
+		return state, err
+	}
+	defer file.Close()
+
+	buf := protocol.BufferPool.Get(state.file.BlockSize())
+	defer protocol.BufferPool.Put(buf)
+
+	finder := weakhash.NewFinder(file, buf)
+
+	byWeakhash := make(map[uint32][]int) // Block indices by weak hash.
+	for i, block := range state.blocks {
+		if h := block.WeakHash; h != 0 {
+			finder.Add(h)
+			byWeakhash[h] = append(byWeakhash[h], i)
+		}
+	}
+
+	if len(byWeakhash) == 0 {
+		l.Debugf("not weak hashing %s. file did not contain any weak hashes", state.file.Name)
+		return state, nil
+	}
+
+	done := make([]bool, len(state.blocks))
+
+	for finder.Next() {
+		h, offset := finder.Match()
+		sha := sha256.Sum256(buf)
```

Comment:
Match() needs to return buf, because now buf is size of BlockSize(), yet last block can be smaller than that.

Reply:
We could return the size of buf, and then before hashing, make sure that the candidate blocks are of the expected size before wasting time doing sha256, but perhaps that's micro-optimization.

Reply:
Good catch. I've added a check in sendReceiveFolder.copyByWeakHash so that the last block is never shown to the weakhash.Finder if it isn't of the right size.

Reply:
Could you point out where/what exactly that check is, I fail to find it.